pkg/fetch/http_fetcher.go

```diff
@@ -90,7 +91,7 @@ func (hf *httpFetcher) FetchDirectory(ctx context.Context, req *remoteasset.Fetc
 }
 
 func (hf *httpFetcher) CheckQualifiers(qualifiers qualifier.Set) qualifier.Set {
-	return qualifier.Difference(qualifiers, qualifier.NewSet([]string{"checksum.sri", "bazel.auth_headers"}))
+	return qualifier.Difference(qualifiers, qualifier.NewSet([]string{"checksum.sri", "bazel.auth_headers", "bazel.canonical_id"}))
```

Comment:
Looking at https://github.com/bazelbuild/proposals/blob/main/designs/2019-04-29-cache.md - it looks like the canonical_id changing should cause a redownload, even if the SHA256 doesn't change? It looks like the bb-remote-asset code serializes the URIs + sorted qualifier list into a proto message and then inserts this into (a) CAS - so changing any qualifier (including the canonical_id) should cause a cache miss and a re-fetch - and thus this qualifier is safe to blindly add to the list with no further code changes - or am I missing something fundamental?

Reply:
Your understanding is correct!  We treat URIs and qualifiers together as a unique key for an asset, and so if any qualifier changes the key used will change, and we will have a cache miss.  At the time we wrote this logic there were no use cases implemented that required partial matches on qualifier sets, so we bundled them all up so that a fetch requires all to match when cached -- as far as I know, this is still the case.